components/tidb_query_datatype/src/codec/mysql/json/binary.rs

```diff
@@ -5,9 +5,26 @@ use std::convert::TryInto;
 use codec::number::NumberCodec;
 
 use super::{constants::*, JsonRef, JsonType, ERR_CONVERT_FAILED};
-use crate::codec::Result;
+use crate::codec::{mysql::json::path_expr::ArrayIndex, Result};
 
 impl<'a> JsonRef<'a> {
+    /// Gets the index from the ArrayIndex
+    ///
+    /// If the idx is from end, and is greater than the count, it will return
+    /// None
+    pub fn array_get_index(&self, idx: ArrayIndex) -> Option<usize> {
```

Comment:
Could you please record the ref to `tidb` as L30 did? I have no idea about the logic with no reference==

Reply:
The parser of json path in TiDB locates in https://github.com/pingcap/tidb/blob/a9373ea526e987396d286878a136501bbcc94d1d/types/json_path_expr.go#L304, and the function `array_get_index` is similar with https://github.com/pingcap/tidb/pull/38584/files#diff-c122a0d90d56f1003723e643f1e0fe9a4f986b964dfa8de985722a4a990ea0eaR58.

There are two difference between TiDB implementation and TiKV:
1. TiDB doesn't have good lightweight parser libraries like `nom`, so it's manually implemented.
2. Go doesn't have tagged union, so we use negative number to represent `[last - xxx]`, use positive number to represent `[xxx]`.

The related PR is https://github.com/pingcap/tidb/pull/38584